### PR TITLE
Add normalize field transformation

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
### Checklist

- [X] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [X] Any modified or new methods or classes have helpful JSDoc and code is thoroughly commented
- [na] The description lists all applicable issues this PR seeks to resolve
- [na] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

This PR adds UI features to normalize field names on processing a new GTFS zip file:
- Capitalize field names, with exceptions
- Substitute text/words with other text (and normalize space if replacing words).

Companion PR: https://github.com/ibi-group/datatools-server/pull/371.

Screenshot (updated):

image

### CODE REVIEW

**Review by evansiroky — CHANGES_REQUESTED** [review]
Before submitting for review, can you please merge `dev` into this branch so that GitHub Actions are ran?

File: lib/manager/components/transform/NormalizeField.js
Comment (evansiroky) [inline lib/manager/components/transform/NormalizeField.js:108]: This method isn't recommended according to a [React blog post](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key). To reset some values, one recommended option is to pass in a `key` prop that will be different whenever the transformation value is different.

File: lib/manager/components/transform/ReplaceFileFromString.js
Comment (evansiroky) [inline lib/manager/components/transform/ReplaceFileFromString.js:41]: This method isn't recommended according to a [React blog post](https://reactjs.org/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key). To reset some values, one recommended option is to pass in a `key` prop that will be different whenever the transformation value is different.

File: lib/manager/components/transform/ReplaceFileFromString.js
Comment (evansiroky) [inline lib/manager/components/transform/ReplaceFileFromString.js:43]: I see that these methods are alphabetized, but I kind of like it when `render` is the last method. It's easier to find that way.

File: lib/manager/components/transform/FeedTransformation.js
Comment (evansiroky) [inline lib/manager/components/transform/FeedTransformation.js:74]: This sends over the `transformation` prop and not the values in the current state of the transformation. So if the user edits the NormalizeFieldTransformation for example, the validation issues won't go away until the transformation is persisted.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See comments, but only the comment about the validation issues is blocking.

File: lib/manager/components/transform/NormalizeField.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/NormalizeField.js]: Good suggestion, updated in https://github.com/ibi-group/datatools-ui/pull/649/commits/a7296121781a1413670f4e5ad6eb4bafabbee7bb.

File: lib/manager/components/transform/ReplaceFileFromString.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/ReplaceFileFromString.js]: Updated in https://github.com/ibi-group/datatools-ui/pull/649/commits/a7296121781a1413670f4e5ad6eb4bafabbee7bb.

File: lib/manager/components/transform/ReplaceFileFromString.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/ReplaceFileFromString.js:47]: Updated in https://github.com/ibi-group/datatools-ui/pull/649/commits/a7296121781a1413670f4e5ad6eb4bafabbee7bb.

File: lib/manager/components/transform/FeedTransformation.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/FeedTransformation.js]: Thanks for pushing this. I moved validation code of the editing state inside each transform component in https://github.com/ibi-group/datatools-ui/pull/649/commits/ccc561232dfc744f574b02b6b35c44cdf0a73f9b.

**Comment by binh-dam-ibigroup** [comment]
Assigning this back to myself to implement substitutions as a structure list.

**Comment by binh-dam-ibigroup** [comment]
Commit https://github.com/ibi-group/datatools-ui/pull/649/commits/67a416c654f106bee6074525549ae7aa1d2e3884 replaces the funky `=>`/`=>*` syntax and now matches the structure requested in https://github.com/ibi-group/datatools-server/pull/371#discussion_r617092209.

File: lib/types/index.js
Comment (evansiroky) [inline lib/types/index.js:324]: I'd prefer if each transformation type extended the `FeedTransformationBase` and then this `FeedTransformation` type would be a [union type](https://flow.org/en/docs/types/unions/) of all the possible transformations.

File: lib/manager/components/transform/Transform.js
Comment (evansiroky) [inline lib/manager/components/transform/Transform.js:21]: [Inheritance is not recommended for creating React components](https://reactjs.org/docs/composition-vs-inheritance.html). In here it seems like the `updateState` is the main way components communicate with this component which then will call `onValidationErrors`. Perhaps one way to go to avoid inheritance is to create a component that handles all of the differences between each transformation type. But by then this all might start to look like the beginning of a form state management component which kind of brings up the question of why aren't we using something like Formik. Also, maybe things would be easier if all these transformRules were a separate mongo collection instead of a list within a feed source record.

**Review by evansiroky — CHANGES_REQUESTED** [review]
This addition of components using inheritance is giving me red flag vibes. Something doesn't seem right here and a decent refactor may be needed to simplify things.

File: lib/manager/components/transform/NormalizeField.js
Comment (landonreed) [inline lib/manager/components/transform/NormalizeField.js:275]: Please place this component in a separate file. This file is too large.

File: lib/manager/components/transform/Transform.js
Comment (landonreed) [reply lib/manager/components/transform/Transform.js:21]: Yea, this seems like a bit of a radical change to the structure of the components that might have benefitted from a technical discussion before taking the plunge.

File: lib/manager/components/transform/FeedTransformation.js
Comment (landonreed) [inline lib/manager/components/transform/FeedTransformation.js:117]: Why make this unsearchable?

File: lib/manager/components/transform/NormalizeField.js
Comment (landonreed) [inline lib/manager/components/transform/NormalizeField.js:258]: I don't think this should be unsearchable.

File: lib/manager/components/transform/NormalizeField.js
Comment (landonreed) [inline lib/manager/components/transform/NormalizeField.js:245]: I could see this getting reused, so it would be nice to place it in its own file.

File: lib/manager/components/transform/ReplaceFileFromString.js
Comment (landonreed) [inline lib/manager/components/transform/ReplaceFileFromString.js:15]: I agree with @evansiroky about avoiding inheritance... it just takes us away from patterns that are used everywhere else in our apps and makes it tough for the reader to switch contexts.

**Review by landonreed — CHANGES_REQUESTED** [review]
See comments about inheritance, making files smaller, and issue saving field selection on QA server (mentioned on Teams).

File: lib/manager/components/transform/NormalizeField.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/NormalizeField.js]: Moved in https://github.com/ibi-group/datatools-ui/pull/649/commits/19fb3d2f7d2a415b941f3b47e667877c0958fe69.

File: lib/manager/components/transform/ReplaceFileFromString.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/ReplaceFileFromString.js:13]: Addressed in https://github.com/ibi-group/datatools-ui/pull/649/commits/54f1aac3573cc5af4575bd839afb116708e626e8

File: lib/manager/components/transform/Transform.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/Transform.js]: I changed that in https://github.com/ibi-group/datatools-ui/pull/649/commits/54f1aac3573cc5af4575bd839afb116708e626e8, although it creates some code duplication for managing the state.

File: lib/types/index.js
Comment (binh-dam-ibigroup) [reply lib/types/index.js]: Updated in https://github.com/ibi-group/datatools-ui/pull/649/commits/54f1aac3573cc5af4575bd839afb116708e626e8. There is one prop `typeName` I had to add that duplicates `@type`, without which flow rejects the case in `getTransformationName` that involves `csvData`. (Flow does not seem to accept attribute `@type` for type branching)

File: lib/manager/components/transform/NormalizeField.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/NormalizeField.js]: This was done in https://github.com/ibi-group/datatools-ui/pull/649/commits/7778d4bb3123d39936be338917e1331f1f93a89b.

File: lib/manager/components/transform/NormalizeField.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/NormalizeField.js]: This is done as part of in https://github.com/ibi-group/datatools-ui/pull/649/commits/7778d4bb3123d39936be338917e1331f1f93a89b.

File: lib/manager/components/transform/FeedTransformation.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/FeedTransformation.js]: Reverted in https://github.com/ibi-group/datatools-ui/pull/649/commits/9a4a01d4b5a48e77d70e5e34c9f20226a04a2e8e.

**Review by evansiroky — APPROVED** [review]
Great stuff, @binh-dam-ibigroup. I think the slightly extra amount of duplicate code vs adding inheritance will keep things simpler and easier to maintain.

File: lib/manager/components/transform/ReplaceFileFromVersion.js
Comment (landonreed) [inline lib/manager/components/transform/ReplaceFileFromVersion.js:51]: Generally, I think the pattern we follow is to place lifecycle methods like this at the beginning of the component. Same comment goes for the other files in this PR.

**Review by landonreed — APPROVED** [review]
Looks great! Approved once #681 is merged.

**Comment by binh-dam-ibigroup** [comment]
Ready for review again. Comment https://github.com/ibi-group/datatools-ui/pull/681#issuecomment-840832999 is addressed by (1) https://github.com/ibi-group/datatools-ui/pull/649/commits/c87759946e6f4063138a115f870aa16d28c5e94a and (2) https://github.com/ibi-group/datatools-ui/pull/649/commits/9f10808f1536bc3c6eac246b38ff77106f211e42 / https://github.com/ibi-group/datatools-server/pull/371/commits/6fb9a0ffb65deeaeb979311be04d70f702813abc.

File: lib/manager/components/transform/NormalizeField.js
Comment (evansiroky) [inline lib/manager/components/transform/NormalizeField.js:59]: Given a new transformation, when this is executed, then the `this.props.transformation.substitutions` is null which causes the array to become `[null]` which leads to a downstream error anywhere that properties on this supposed expected substitution object are checked. This results in the Add button not adding any substitutions and just error'ing out.

**Review by evansiroky — CHANGES_REQUESTED** [review]
See comment.

File: lib/manager/components/transform/NormalizeField.js
Comment (binh-dam-ibigroup) [reply lib/manager/components/transform/NormalizeField.js]: Thanks for finding this bug! Fixed in. https://github.com/ibi-group/datatools-ui/pull/649/commits/a595d1a6b4f0f173dc330fb9987f0489e49d20ee.

**Review by evansiroky — CHANGES_REQUESTED** [review]
Some oddities I have observed with adding substitutions:

- The placeholders seem to be the actual values upon adding a new rows
- Sometimes if you click "Add" (substitution) with just one row immediately after editing the replacement without blurring the replacement field, no new row is created
- If you click "Add" with more than one row after about 1 second of editing a replacement (without blurring), a new row will be added but the change made to the replacement will be lost

I believe all of these are a result of dealing with a list that is stored within a mongo object. This result in concurrency issues because whenever this list is attempted to be updated, updates in parallel cannot occur otherwise a race condition will exist where one version of the list will win and the updates from the other will be lost.

The solution to this problem could be to either lock editing to only one specific list item at a time or to make the list a new collection within the database that has its own IDs that would enable asynchronous editing.

However, thinking further, this really doesn't do anything about the use case of two users simultaneously editing the same list. But at the very least, I can't approve until at the very least one of the above two solutions is implemented to at least solve the asynchronous editing problem for a single user editing the transformation.

**Comment by binh-dam-ibigroup** [comment]
> * The placeholders seem to be the actual values upon adding a new rows
Correct, they are templates as written. Should they be left blank?

> * Sometimes if you click "Add" (substitution) with just one row immediately after editing the replacement without blurring the replacement field, no new row is created
> * If you click "Add" with more than one row after about 1 second of editing a replacement (without blurring), a new row will be added but the change made to the replacement will be lost
> 
> I believe all of these are a result of dealing with a list that is stored within a mongo object. This result in concurrency issues because whenever this list is attempted to be updated, updates in parallel cannot occur otherwise a race condition will exist where one version of the list will win and the updates from the other will be lost.
As written, there is indeed a race condition, but the behavior you observed comes from the full remounting/rerendering of `NormalizeField` after an edit is made and the persisted values are refetched from the backend. I changed the UI component in https://github.com/ibi-group/datatools-ui/pull/649/commits/6e74f4859f36769af29febb7c4cd3cea96f5475a to queue the add/delete operations, so they execute after the edits take effect.

**Review by evansiroky — CHANGES_REQUESTED** [review]
This definitely will not be able to properly apply updates if there is any noticeable amount of latency with the network connection and a user who edits things really rapidly. Try setting the [network throttling to "fast 3G" in Chrome Dev tools](https://developer.chrome.com/docs/devtools/network/reference/#throttling) and then edit a few different fields as if you were a power-editor. I really recommend implementing a change that would lock an entire record and have a edit/save button which at the very least will slow down the user editing to make sure we aren't sending over data to be saved that isn't built on top of the latest version of that record in the database. I kind of did this when implementing the [custom file editor](https://github.com/ibi-group/datatools-ui/blob/dev/lib/manager/components/deployment/CustomFileEditor.js) in the deployment.

**Comment by binh-dam-ibigroup** [comment]
> This definitely will not be able to properly apply updates if there is any noticeable amount of latency with the network connection and a user who edits things really rapidly. Try setting the [network throttling to "fast 3G" in Chrome Dev tools](https://developer.chrome.com/docs/devtools/network/reference/#throttling) and then edit a few different fields as if you were a power-editor. I really recommend implementing a change that would lock an entire record and have a edit/save button which at the very least will slow down the user editing to make sure we aren't sending over data to be saved that isn't built on top of the latest version of that record in the database. I kind of did this when implementing the [custom file editor](https://github.com/ibi-group/datatools-ui/blob/dev/lib/manager/components/deployment/CustomFileEditor.js) in the deployment.

@evansiroky Save/revert functionality was available until https://github.com/ibi-group/datatools-ui/pull/649/commits/c87759946e6f4063138a115f870aa16d28c5e94a by suggestion of @landonreed. Given the added unnecessary network traffic inline edits create, I am in favor of going back to a Save/revert functionality. But then, for consistency, table selector should also be converted to work that way, so there is a potentially large refactor. Thoughts?

**Comment by binh-dam-ibigroup** [comment]
I made changes as discussed during our call, see commits https://github.com/ibi-group/datatools-ui/pull/649/commits/1d483020b1a4314b3c85f1a755437c95fc476f50 and following.

**Review by landonreed — CHANGES_REQUESTED** [review]
@binh-dam-ibigroup this looks really good! A couple of tweaks on the UX:
- it's not obvious that you can click a row to begin editing. It might help users know this is ok by adding `cursor: pointer` to the style when "click to edit" is allowed.
- We should probably just prohibit users from saving a row if it is invalid. Right now, I can clear the search text and click "save", but only then am I shown the validation issue.
- when I click "Add", I kind of expect the added row to be in the "editable" state, rather than having to also click "Edit".

I haven't looked at the code changes yet, but wanted to get your eyes on these first.

**Comment by binh-dam-ibigroup** [comment]
> * We should probably just prohibit users from saving a row if it is invalid. Right now, I can clear the search text and click "save", but only then am I shown the validation issue.
Good idea.

> * when I click "Add", I kind of expect the added row to be in the "editable" state, rather than having to also click "Edit".
> 
I initially felt the same about it. One side effect: that action will prevent the user to add two new rows, they would have to save or dismiss a newly created row first before they can add a new one. Do we want that?

**Comment by landonreed** [comment]
> One side effect: that action will prevent the user to add two new rows, they would have to save or dismiss a newly created row first before they can add a new one. Do we want that?

Yes, I think that side effect is fine. Also, I'm not really sure it's valuable to pre-populate the value of the search/replace with fields.

**Comment by binh-dam-ibigroup** [comment]
Ready for review again, see the latest tweaks from https://github.com/ibi-group/datatools-ui/pull/649/commits/5839fc9a13a9587531d57306d585e207086374c9.

**Review by evansiroky — APPROVED** [review]
Approving so we can move along, but I still think there are 2 design problems that we should seriously consider avoiding in the future.

1. The decision to make an a List of individually-editable entities within a record of another parent object. We should seriously consider avoiding creating a list field within a record unless all entities must be edited at once. We should also consider refactoring the existing cases where this occurs.
2. The implementation of a custom form made with code. We should consider sticking to a form library like Formik that will reduce a lot of this boilerplate. Furthermore, using a single form library will standardize the look and feel of other places that have data entry elsewhere in the application. In this PR we went back and forth a lot before finally coming to a decision on how to create the inputs. If we had used a standard single form library, there would be an expectation of how to implement form inputs and therefore the coding and review time should be shorter.

**Review by landonreed — CHANGES_REQUESTED** [review]
- I think things would be easier/cleaner if we defaulted the substitutions list to an empty list (rather than null)
- There's a minor bug in the ReplaceFileFromVersion transformation causing a validation issue to show when it is not relevant (doesn't appear to be a problem with ReplaceFileFromString):
image

**Comment by binh-dam-ibigroup** [comment]
> * There's a minor bug in the ReplaceFileFromVersion transformation causing a validation issue to show when it is not relevant (doesn't appear to be a problem with ReplaceFileFromString):
>   image

Good find, this is fixed in https://github.com/ibi-group/datatools-ui/pull/649/commits/7b00cee840ea112c6961167dc26b613ec040ddc8.

**Comment by binh-dam-ibigroup** [comment]
> * I think things would be easier/cleaner if we defaulted the substitutions list to an empty list (rather than null)

That comes from the backend hence the null check. I refactored that null check in https://github.com/ibi-group/datatools-ui/pull/649/commits/fadcc23808f74c0321112eb6997242c900ed665a.

**Comment by binh-dam-ibigroup** [comment]
The latest changes from https://github.com/ibi-group/datatools-ui/pull/649/commits/7897565431e26ddcb393990c0b00951fac42a6ac require https://github.com/ibi-group/datatools-server/pull/386.

**Review by landonreed — APPROVED** [review]

**Pull request merged**
